strawberryfields/backends/bosonicbackend/bosoniccircuit.py

```diff
@@ -0,0 +1,833 @@
+                )
+
+                exp_arg = np.einsum(
+                    "...j,...jk,...k",
+                    (peak_sample - self.means[:, expind]),
+                    np.linalg.inv(self.covs[:, expind, :][:, :, expind].real + covmat),
+                    (peak_sample - self.means[:, expind]),
+                )
+                ub_exp_arg = np.copy(exp_arg)
+                if imag_means_ind.size > 0:
+                    ub_exp_arg[imag_means_ind] = np.einsum(
+                        "...j,...jk,...k",
+                        (peak_sample - self.means[imag_means_ind, :][:, expind].real),
+                        np.linalg.inv(
+                            self.covs[imag_means_ind, :, :][:, expind, :][:, :, expind].real
+                            + covmat
+                        ),
+                        (peak_sample - self.means[imag_means_ind, :][:, expind].real),
+                    )
+                prob_dist_val = np.real_if_close(
+                    np.sum(
+                        (
+                            np.array(self.weights)
+                            / np.sqrt(
+                                np.linalg.det(
+                                    2
+                                    * np.pi
+                                    * (self.covs[:, expind, :][:, :, expind].real + covmat)
+                                )
+                            )
+                        )
+                        * np.exp(-0.5 * exp_arg)
+                    )
+                )
+                prob_upbnd = np.real_if_close(
+                    np.sum(
+                        (
+                            ub_weights
+                            / np.sqrt(
+                                np.linalg.det(
+                                    2
+                                    * np.pi
+                                    * (
+                                        self.covs[ub_ind, :, :][:, expind, :][:, :, expind].real
+                                        + covmat
+                                    )
+                                )
+                            )
+                        )
+                        * np.exp(-0.5 * ub_exp_arg[ub_ind])
+                    )
+                )
+                vertical_sample = np.random.random(size=1) * prob_upbnd
+                if vertical_sample < prob_dist_val:
+                    drawn = True
+                    vals[i] = peak_sample
+        # The next line is a hack in that it only updates conditioned on the first samples value
+        # should still work if shots = 1
+        if len(indices) < len(self.active):
+            self.post_select_generaldyne(covmat, indices, vals[0])
+
+        # If all modes are measured, set them to vacuum
+        if len(indices) == len(self.active):
+            for i in indices:
+                self.loss(0, i)
+
+        return vals
+
+    def homodyne(self, n, shots=1, eps=0.0002):
```

Comment:
Any pointers on the choosing the default value of `eps`?

Reply:
A homodyne measurement is a projection onto an eigenstate of the quadrature, which can be approximated by taking a state that's super-squeezed in that quadrature, hence setting eps to be so small.